### DISCUSSION TITLE
How to enforce ranges or other data conditions?

### DISCUSSION THREAD

**Discussion by JiveyGuy** [jam1garner/binrw#214] [Q&A] [answered]
If I have a struct with several u8 and I want to panic during reading if one of the values is outside an arbitrary range (i, j), how would I accomplish this? 

More generally how can we enforce data meets any criteria?

**Comment by csnover** [comment, marked as answer]
The [assert](https://docs.rs/binrw/latest/binrw/docs/attribute/index.html#assert) directive validates objects and fields after they are read or before they are written, returning an error if the assertion condition evaluates to false. Any earlier field or [import](https://docs.rs/binrw/latest/binrw/docs/attribute/index.html#arguments) can be referenced by expressions in the directive.

For this specific case you would assert on `#[br(assert((i..j).contains(field)))]`.

**Reply by JiveyGuy** [reply]
Amazing! Thanks!
